ci: install perl provider

janlazo opened the pull request:
Resolve CI dependencies for https://github.com/neovim/neovim/pull/11595

jacquesg committed:
remote plugins: add support for perl hosts

janlazo committed:
ci: install perl provider

justinmk commented:
travis macOS job failed:

```
--> Working on MsgPack::Raw
Fetching http://www.cpan.org/authors/id/J/JA/JACQUESG/MsgPack-Raw-0.05.tar.gz ... OK
Configuring MsgPack-Raw-0.05 ... OK
Building MsgPack-Raw-0.05 ... FAIL
! Installing MsgPack::Raw failed. See /Users/travis/.cpanm/work/1578775862.7416/build.log for details. Retry with --force to force install it.
! Installing the dependencies failed: Module 'MsgPack::Raw' is not installed
! Bailing out the installation for Neovim-Ext-0.02.
```

janlazo commented:
Is it the system perl or something else? I'll try with the perl homebrew package. Else, use perlbrew or https://github.com/tokuhirom/plenv.

janlazo commented on ci/before_build.ps1:
```diff
@@ -0,0 +1,8 @@
+# https://github.com/lukesampson/scoop#installation
+Invoke-Expression (New-Object System.Net.WebClient).DownloadString('https://get.scoop.sh')
+
+Set-StrictMode -Version Latest
```
If this doesn't work, scoop can't be used in the CI without disabling strict mode.

janlazo replied:
It works.

janlazo commented on ci/build.ps1:
```diff
@@ -57,7 +57,7 @@ if ($compiler -eq 'MINGW') {
   $cmakeGenerator = 'Ninja'
   $cmakeGeneratorArgs = '-v'
-  $mingwPackages = @('ninja', 'cmake', 'perl', 'diffutils').ForEach({
+  $mingwPackages = @('ninja', 'cmake', 'diffutils').ForEach({
```
Why is perl needed here again? It should safe to skip it if `C:/perl` in Appveyor is from http://strawberryperl.com/.

justinmk replied:
perl is needed for git, maybe other things. Though I suppose we can try this...

janlazo commented on .travis.yml:
```diff
@@ -90,7 +91,9 @@ addons:
       - ccache
+      - cpanminus
       - ninja
+      - perl
```
osx job passed so it's safe to assume that the perl provider supports recent releases only for osx.

jacquesg replied:
It works at least all the way back to perl 5.14, but should work for perl 5.12 and 5.10 as well:

https://dev.azure.com/jacquesgermishuys/p5-Neovim-Ext/_build/results?buildId=183&view=results

janlazo replied:
Can you investigate this in your Azure pipeline by creating a new job that doesn't use perlbrew?

jacquesg replied:
I've tested this locally. Seems like the required perl headers are missing by default on macOS. They would need to be installed to use the system perl.

https://stackoverflow.com/questions/52682304/fatal-error-extern-h-file-not-found-while-installing-perl-modules
https://developer.apple.com/documentation/xcode_release_notes/xcode_10_release_notes#3035624

IMHO I think using perl from homebrew would be the simplest.

janlazo committed:
ci/Appveyor: install diffutils via scoop

scoop is a Windows package manager, similar to homebrew on OSX.
It is written on powershell, works in unelevated accounts,
is regularly maintained, and does not need mingw/msys2/cygwin
or WSL environments.
Sample use case is running (busybox) bash,
which can be installed via scoop via "scoop install busybox",
to use CI bash scripts in Appveyor.

Parts of scoop's directory can be cached to avoid downloads.
scoop modifies the registry for persistence environment variables
so scoop itself cannot be cached.

janlazo commented:
Now that perl provider can be installed in Windows,OSX,Linux, how can it be tested?

justinmk commented:
> Now that perl provider can be installed in Windows,OSX,Linux, how can it be tested?

I assume after that, next step would be https://github.com/neovim/neovim/pull/11595 . Then perhaps someone wants to fill in `ex_perl` and friends? Here's how it was done for ruby: 0a3c0205c5b9043d371133

justinmk commented on ci/build.ps1:
```diff
@@ -116,6 +130,8 @@ if (-not $NoTests) {
   }
 
+  cpanm.bat -n Neovim::Ext
+  perl -W -e 'use Neovim::Ext; print $Neovim::Ext::VERSION'
```
perhaps this block should live near the npm/ruby stuff (line ~116)

janlazo replied:
It's the same install order as in `ci/install.sh`. I don't want to move it up because perl modules are often compiled. If anything, I may move it down if I need more OS packages.

janlazo commented:
> could you summarize (in commit message, ideally) why scoop is a better option than e.g. powershell `Install-Product` (for node at least, if not perl).

scoop is like homebrew on Windows with the differences that it uses powershell and runs on Windows instead of using ruby and running on OSX. scoop is easy to install and manage compared to alternatives (no need for mingw/msys2/cygwin unlike pacman, works with non-admin account unlike chocolately, `Install-Product` is specific to Appveyor). scoop is regularly updated so there's less need to tinker with the package versions (no need to bump node versions anymore).

Cons would be caching and registry modification for `PATH` because of lack of integration with Appveyor. I don't know how to decouple scoop installation from the build script. It's fine for now because we're using 1 powershell script for the build and scoop is fast enough.

janlazo committed:
provider/perl: add basic tests

janlazo commented on test/functional/provider/perl_spec.lua:
```diff
@@ -0,0 +1,42 @@
+
+      my $session = Neovim::Ext::MsgPack::RPC::socket_session($ENV{NVIM_LISTEN_ADDRESS});
+      my $nvim = Neovim::Ext::from_session($session);
+      $nvim->command('let g:job_out = "hello"');
```
@jacquesg What's the simplest plugin abstraction that does this within one file (no `rplugin/perl/`)? I can't figure out how to use the Plugin class for this without creating a new file. For reference, I want to do the equivalent of what's in the nodejs_spec.lua.

janlazo replied:
Nevermind. I found a simple workaround.

janlazo commented on test/functional/provider/perl_spec.lua:
```diff
@@ -0,0 +1,73 @@
+      use parent qw(Neovim::Ext::Plugin);
+
+      @{TestPlugin::commands} = ();
+      @{TestPlugin::specs} = ();
```
@jacquesg I had to add these for perl v5.22.1 (Ubuntu Xenial) to register the plugin without warnings. Can you fix the warnings?

jacquesg replied:
What are the warnings you are seeing?

janlazo replied:
`<Symbol> used only once: possible typo at <file> line <line number>`. The decorator sets these to empty arrays but perl complains that these properties aren't used.

janlazo commented on test/functional/provider/perl_spec.lua:
```diff
@@ -0,0 +1,73 @@
+      my $session = Neovim::Ext::MsgPack::RPC::socket_session($ENV{NVIM_LISTEN_ADDRESS});
+      my $nvim = Neovim::Ext::from_session($session);
+      my $plugin = Neovim::Ext::Plugin::new('TestPlugin', $nvim);
+      $plugin->test_command();
```
This should run `TestCommand` but it needs the Host class to start the event loop. I don't know how that works.

jacquesg replied:
A plugin should derive from ```Neovim::Ext::Plugin```

https://github.com/jacquesg/p5-Neovim-Ext/blob/master/t/rplugin/perl/TestPlugin.pm

To start the host could be as simple as [start_host()](https://github.com/jacquesg/p5-Neovim-Ext/blob/master/lib/Neovim/Ext.pm#L59)

janlazo replied:
Did I not do that for this test? I used `package` twice, one for the plugin and one for the session. I tried `start_host` before but it assumes `rplugin/perl` in `rtp`. I'm not testing that yet.

jacquesg replied:
The syntax to instaniate the plugin isn't quite right. It should eithe be:
```my $plugin = TestPlugin->new ($nvim)```
or
```my $plugin = new TestPlugin ($nvim)```;

https://github.com/jacquesg/p5-Neovim-Ext/blob/master/lib/Neovim/Ext/Plugin.pm#L202

It should also register the plugin like so:
https://github.com/jacquesg/p5-Neovim-Ext/blob/master/t/rplugin/perl/TestPlugin.pm#L7

janlazo commented:
Needs a health check for perl provider and this should ready to try out before proceeding with `:perl`.

janlazo committed:
provider/perl: simplify detection

janlazo commented:
Wonder why MSVC builds don't work or is it just 32-bit builds in general because cpan modules are compiled?

janlazo commented on runtime/autoload/provider/perl.vim:
```diff
@@ -0,0 +1,69 @@
+  endif
+
+  " if perl is available, make sure the required module is available
+  call system([prog, '-W', '-MNeovim::Ext', '-e', ''])
```
Used `system` because this is quick. Could use `jobstart` for https://metacpan.org/pod/local::lib, `perlbrew`, https://github.com/tokuhirom/plenv.

janlazo committed:
provider/perl: add health check

janlazo commented on runtime/autoload/health/provider.vim:
```diff
@@ -623,10 +623,51 @@ function! s:check_node() abort
+          \ ['Install Perl and cpanminus and verify that `perl` and `cpanm` commands work.'])
+    return
+  endif
+  let perl_v = get(split(s:system(['perl', '-W', '-e', 'print $^V']), "\n"), 0, '')
```
Require Perl v5.6.0.

https://perldoc.perl.org/perlvar.html#%24%5eV

janlazo commented on runtime/autoload/health/provider.vim:
```diff
@@ -623,10 +623,51 @@ function! s:check_node() abort
+    return
+  endif
+
+  call health#report_ok('"Neovim::Ext" cpan package is installed: '. current_perl)
```
@jacquesg Is there a way to query cpan for latest version of Neovim::Ext?

jacquesg replied:
You could use ```cpanm --info Neovim::Ext``` which would produce something like:

```JACQUESG/Neovim-Ext-0.02.tar.gz```

janlazo replied:
It works but I can't tell if that version check is for locally installed modules or from CPAN.

jacquesg replied:
It is from CPAN, tested it on a vanilla perl.

janlazo commented:
https://ci.appveyor.com/project/neovim/neovim/builds/30204102/job/8kiqaw62blm0mhtu
```
Building and testing IO-Async-0.75
Building IO-Async
t\00use.t .................... ok
t\01timequeue.t .............. ok
t\02os.t ..................... ok
t\03loop-magic.t ............. ok
t\04notifier.t ............... ok
t\05notifier-loop.t .......... ok
t\06notifier-mixin.t ......... ok
t\07notifier-future.t ........ ok
#   Failed test 'exceptional socket invokes on_read_ready'
#   at C:\Users\appveyor\.cpanm\work\1579357430.3792\IO-Async-0.75\blib\lib/IO/Async/LoopTests.pm line 390.
#          got: '0'
#     expected: '1'
# Looks like you failed 1 test of 22.
t\10loop-poll-io.t ........... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/22 subtests 
	(less 2 skipped subtests: 19 okay)
```

janlazo committed:
provider/perl: skip tests on Windows

It worked for MINGW builds at one point
but it keeps failing now because of perl dependencies
or nvim session issues for tests (named pipes as sockets on Windows?).

janlazo committed:
doc: provider-perl

Enable 'has("perl")' as an alias for 'g:loaded_perl_provider'.

TODO:
- +perl interface

janlazo commented on ci/install.sh:
```diff
@@ -50,3 +50,6 @@ else
 fi
 test -f "$TREE_SITTER_DIR/bin/c.so"
+
+sudo cpanm -n Neovim::Ext || cat "$HOME/.cpanm/build.log"
```
This fails when running `Neovim::Ext` tests because it can't download nvim v0.4.3.

jacquesg replied:
It should only download nvim if it is not on the path, and when running tests. ```-n``` here should skip the tests?

janlazo replied:
The issue is it can't download nvim when cpanm verified that the required Http perl modules are installed. Neovim::Ext dependencies don't fail their tests so `-n` is only for `Neovim::Ext` tests.

jacquesg replied:
```-n``` should apply to the distribution you're installing and all of its dependencies, if it doesn't, then I believe its a bug in ```App::cpanminus```. ```Neovim::Ext``` downloads neovim to verify that the API actually works. Not sure how the HTTP perl modules is in play here?

janlazo replied:
https://github.com/jacquesg/p5-Neovim-Ext/commit/82565d72a8659ebca9196cd672e44f6c7f436b01#diff-8424792320c9483850c5842f19129c9bR89-R94

jacquesg replied:
I don't get what your point is? Like I said above, yes, ```Neovim::Ext``` downloads nvim to run its test if its not on the ```PATH```. However, running ```cpanm``` with the ```-n```/```--notest``` switch will skip tests and the download of neovim all together.

janlazo committed:
provider/perl: add latest version health check

janlazo committed:
provider/perl: test older versions

justinmk commented on ci/build.ps1:
```diff
@@ -29,6 +29,26 @@ function exitIfFailed() {
+  Get-Content -Path "$env:USERPROFILE\.cpanm\build.log"
+}
+perl -W -e 'use Neovim::Ext; print $Neovim::Ext::VERSION'; exitIfFailed
+
```
this new block should be guarded by

    if (-not $NoTests) {

else it defeats use cases like https://github.com/neovim/neovim/pull/11654

justinmk reviewed (approved):
LGTM after small comment re build.ps1 https://github.com/neovim/neovim/pull/11703#discussion_r368765525

janlazo committed:
ci/Appveyor: respect -NoTest param

Skips nodejs and Neovim::Ext.

janlazo commented on test/functional/provider/perl_spec.lua:
```diff
@@ -0,0 +1,80 @@
+end)
+
+describe('perl host', function()
+  if helpers.pending_win32(pending) then return end
```
https://github.com/neovim/neovim/pull/11595#discussion_r361154077
stdin polling shouldn't be a blocker for these tests on Windows since the host is not running but these tests always fail. My guess is the way I used `socket_session` to create a new session.

justinmk commented on ci/before_install.sh:
```diff
@@ -44,7 +44,6 @@ fi
 
 source ~/.nvm/nvm.sh
 nvm install 10
```
Future idea: perhaps only the "gcov" job should install ruby/python/nodejs/perl. That will make CI (1) more reliable and (2) perhaps save 30-60 seconds per job.
